### PR TITLE
fix: validate configuration and vectorizer loading

### PR DESCRIPTION
**Pull request by hellkonig**
## Summary
- add shared supported log-level validation for config, CLI, and logging setup
- enforce scalar config constraints for top_k and cache_size
- reject unsupported log levels instead of silently falling back to INFO
- add clearer vectorizer loading errors for missing modules, missing classes, missing dependencies, invalid classes, and model instantiation failures

## Issue
Part of #28.

This PR covers the config and vectorizer-loading validation portion. The paper metadata and typed recommendation model work should follow in a separate PR.

## Verification
- uv run ruff check .
- uv run ruff format --check .
- uv run python -m mypy src
- uv run python -m pytest

### CODE REVIEW

File: src/arxiv_recommender/utils/model_loader.py [hunk 1]
```diff
@@ -23,20 +29,79 @@ def load_vectorization_model(
...
+            logging.error("Vectorizer module '%s' was not found.", full_module_name)
+            raise VectorizationModelLoadError(
+                f"Vectorizer module '{full_module_name}' was not found."
+            ) from exc
+
+        logging.error(
+            "Vectorizer module '%s' could not be imported because dependency '%s' was not found.",
+            full_module_name,
+            exc.name,
+        )
```
Comment (hellkonig) [inline src/arxiv_recommender/utils/model_loader.py:49]: can you explain in which case `exc.name == full_module_name`, and why do you need this block?

File: tests/schemas/test_config.py [hunk 2]
```diff
@@ -115,7 +135,41 @@ def test_app_config_from_json_dict(self):
...
         assert config.vectorizer.model_name == "bert-base"
 
-    def test_app_config_required_fields(self):
+    def test_app_config_required_fields(self) -> None:
         """Test that required fields are enforced."""
         with pytest.raises(Exception):
-            AppConfig()
+            AppConfig.model_validate({})
+
+    @pytest.mark.parametrize("top_k", [0, -1])
```
Comment (hellkonig) [inline tests/schemas/test_config.py:143]: Can you explain the syntax here? I think the value of `top_k` should be `-1`, but why is it here `[0, -1]`?

File: tests/utils/test_logging.py [hunk 3]
```diff
@@ -69,6 +69,18 @@ def test_setup_logging_suppresses_third_party(self) -> None:
...
+        """Test that unsupported log levels raise a clear error."""
+        with self.assertRaises(ValueError):
+            setup_logging(level="LOUD", json_format=False)
+
+    def test_supported_log_levels_are_ordered(self) -> None:
+        """Test supported log levels remain deterministic for CLI choices and messages."""
+        self.assertEqual(
+            SUPPORTED_LOG_LEVELS,
+            ("DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"),
+        )
```
Comment (hellkonig) [inline tests/utils/test_logging.py]: Should we hard-code the `SUPPORTED_LOG_LEVELS` here

File: tests/utils/test_model_loader.py [hunk 4]
```diff
@@ -30,73 +49,86 @@ def test_load_default_vectorizer(self, mock_import: Any) -> None:
...
 
-        # Ensure the correct module was loaded
         mock_import.assert_called_once_with("arxiv_recommender.text_vectorization.distil_bert")
-        # Ensure the correct class was instantiated with cache_size
-        mock_class.assert_called_once_with("distilbert-base-uncased", cache_size=1000)
-
-        # Ensure instance is returned
-        self.assertEqual(vectorizer, fake_embedder)
+        self.assertIsInstance(vectorizer, FakeEmbedder)
+        fake_vectorizer = cast(FakeEmbedder, vectorizer)
```
Comment (hellkonig) [inline tests/utils/test_model_loader.py:61]: Can you explain why we need the `cast` here?

File: tests/utils/test_model_loader.py [hunk 5]
```diff
@@ -30,73 +49,86 @@ def test_load_default_vectorizer(self, mock_import: Any) -> None:
...
     def test_load_custom_vectorizer(self, mock_import: Any) -> None:
         """
         Test loading a custom text vectorization model dynamically.
         """
-        mock_module = unittest.mock.MagicMock()
-        mock_class = unittest.mock.MagicMock()
-        fake_embedder = FakeEmbedder()
-        mock_class.return_value = fake_embedder
-        mock_module.CustomVectorizer = mock_class
+        mock_module = SimpleNamespace(CustomVectorizer=FakeEmbedder)
```
Comment (hellkonig) [inline tests/utils/test_model_loader.py:70]: Not sure which part of the code is testing here

File: src/arxiv_recommender/utils/model_loader.py [hunk 6]
```diff
@@ -23,20 +29,79 @@ def load_vectorization_model(
...
+    full_module_name: str,
+) -> Any:
+    model_class = getattr(module, class_name, None)
+    if model_class is None:
+        logging.error("Vectorizer class '%s' was not found in '%s'.", class_name, full_module_name)
+        raise VectorizationModelLoadError(
+            f"Vectorizer class '{class_name}' was not found in module '{full_module_name}'."
+        )
+
+    if not isinstance(model_class, type) or not issubclass(model_class, TextEmbedder):
```
Comment (hellkonig) [inline src/arxiv_recommender/utils/model_loader.py:97]: what does `isinstance(model_class, type)` do here? What is the value of `type`?

File: tests/schemas/test_config.py (hunk 2 above)
Comment (hellkonig) [reply tests/schemas/test_config.py:143]: pytest.mark.parametrize("top_k", [0, -1]) runs this test twice: once with top_k=0, then once with top_k=-1. Both are important boundary cases because the schema requires top_k > 0: zero is the boundary value and -1 represents negative inputs.

File: tests/utils/test_model_loader.py (hunk 4 above)
Comment (hellkonig) [reply tests/utils/test_model_loader.py:61]: load_vectorization_model() intentionally returns the abstract TextEmbedder type. This test then checks attributes defined only on FakeEmbedder. cast(FakeEmbedder, vectorizer) tells mypy about the narrower type after the runtime assertIsInstance() check. It has no runtime behavior.

File: src/arxiv_recommender/utils/model_loader.py (hunk 6 above)
Comment (hellkonig) [reply src/arxiv_recommender/utils/model_loader.py:97]: type is Python built-in metaclass for normal classes. isinstance(FakeEmbedder, type) is true, while isinstance(FakeEmbedder(), type) is false. We check this before calling issubclass(model_class, TextEmbedder) because issubclass() raises TypeError if the configured attribute is not actually a class.

File: tests/utils/test_model_loader.py (hunk 5 above)
Comment (hellkonig) [reply tests/utils/test_model_loader.py:70]: This successful-path test verifies that the expected module name is imported, the configured class is selected from that module, the class is instantiated, model_name and cache_size reach its constructor, and the returned object is a TextEmbedder. The following tests cover each failure branch separately.

File: tests/utils/test_logging.py (hunk 3 above)
Comment (hellkonig) [reply tests/utils/test_logging.py]: Agreed. This test duplicated the implementation tuple and did not add meaningful behavioral coverage. I removed it in commit c497e9e. The behavioral test that unsupported levels are rejected remains.

File: src/arxiv_recommender/utils/model_loader.py (hunk 1 above)
Comment (hellkonig) [reply src/arxiv_recommender/utils/model_loader.py:49]: This branch distinguishes two failure modes. When the configured module itself does not exist, importing arxiv_recommender.text_vectorization.unknown raises ModuleNotFoundError with exc.name equal to full_module_name; the user should fix module_name. If that module exists but imports a missing dependency such as transformers, then exc.name is transformers; the user should fix the environment. The separate block preserves that actionable distinction.
